Possible race condition between selector_ioloop and writer

In our application, we have implemented a client library for both publishing and consuming based on merged asynchronous_publisher_example.py and asynchronous_consumer_example.py.  We have recently encountered errors when publishing a message that appear to be due to a race condition.  It is extremely hard to reproduce - we have only seen it on 3 different processes out of 18, and doesn't always happen on those.

I enabled the logger with an enhanced formatter that includes timestamp and thread ID.  The output below is one example of the issue (others are similar).

```
PIKA: 2020-10-08 17:59:31,330 - 47455081346816 - pika.adapters.utils.selector_ioloop_adapter:313 - DEBUG - SelectorIOServicesAdapter.set_writer(28, <bound method _AsyncPlaintextTransport._on_socket_writable of <pika.adapters.utils.io_services_utils._AsyncPlaintextTransport object at 0x2b28fff2ba10>>)
PIKA: 2020-10-08 17:59:31,331 - 47455081346816 - pika.adapters.utils.selector_ioloop_adapter:330 - DEBUG - set_writer(28, _) updated handler RdWr
PIKA: 2020-10-08 17:59:31,331 - 47455081346816 - pika.adapters.utils.io_services_utils:1023 - DEBUG - Turned on writability watcher: <socket._socketobject object at 0x2b28ffadd130>

## Handle received message here

PIKA: 2020-10-08 17:59:31,334 - 47455081346816 - pika.adapters.utils.io_services_utils:1063 - DEBUG - Recv would block on <socket._socketobject object at 0x2b28ffadd130>
PIKA: 2020-10-08 17:59:31,334 - 47455081346816 - pika.adapters.utils.selector_ioloop_adapter:341 - DEBUG - SelectorIOServicesAdapter.remove_writer(28)
PIKA: 2020-10-08 17:59:31,334 - 47455081346816 - pika.adapters.utils.selector_ioloop_adapter:364 - DEBUG - remove_writer(28) updated handler Rd
PIKA: 2020-10-08 17:59:31,334 - 47455081346816 - pika.adapters.utils.io_services_utils:1112 - DEBUG - Turned off writability watcher: <socket._socketobject object at 0x2b28ffadd130>

## Call channel.basic_publish() to publish a message

PIKA: 2020-10-08 17:59:35,403 - 47455699732224 - pika.adapters.utils.selector_ioloop_adapter:313 - DEBUG - SelectorIOServicesAdapter.set_writer(28, <bound method _AsyncPlaintextTransport._on_socket_writable of <pika.adapters.utils.io_services_utils._AsyncPlaintextTransport object at 0x2b28fff2ba10>>)
PIKA: 2020-10-08 17:59:35,403 - 47455699732224 - pika.adapters.utils.selector_ioloop_adapter:330 - DEBUG - set_writer(28, _) updated handler RdWr
PIKA: 2020-10-08 17:59:35,404 - 47455081346816 - pika.adapters.utils.selector_ioloop_adapter:395 - WARNING - WRITE indicated on fd=28, but writer callback is None; events=0b100
PIKA: 2020-10-08 17:59:35,404 - 47455699732224 - pika.adapters.utils.io_services_utils:1023 - DEBUG - Turned on writability watcher: <socket._socketobject object at 0x2b28ffadd130>
PIKA: 2020-10-08 17:59:35,404 - 47455081346816 - pika.adapters.utils.io_services_utils:57 - ERROR - Wrapped func exited with exception. Caller's stack:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/pika/diagnostic_utils.py", line 53, in log_exception_func_wrap
    return func(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/pika/adapters/utils/io_services_utils.py", line 1093, in _on_socket_writable
    'but _tx_buffers is empty.', self._state)
AssertionError: ('_AsyncPlaintextTransport._on_socket_writable() called, but _tx_buffers is empty.', 1)
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/pika/diagnostic_utils.py", line 53, in log_exception_func_wrap
    return func(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/pika/adapters/utils/io_services_utils.py", line 1093, in _on_socket_writable
    'but _tx_buffers is empty.', self._state)
AssertionError: ('_AsyncPlaintextTransport._on_socket_writable() called, but _tx_buffers is empty.', 1)
PIKA: 2020-10-08 17:59:35,405 - 47455081346816 - pika.adapters.select_connection:807 - DEBUG - Deactivating poller
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/aim/AsyncMsgManager.py", line 604, in run
    self._connection.ioloop.start()
  File "/usr/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 543, in start
    self._poller.start()
  File "/usr/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 803, in start
    self.poll()
  File "/usr/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 1200, in poll
    self._dispatch_fd_events(fd_event_map)
  File "/usr/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 904, in _dispatch_fd_events
    handler(fileno, events)
  File "/usr/lib/python2.7/site-packages/pika/adapters/utils/selector_ioloop_adapter.py", line 391, in _on_reader_writer_fd_events
    callbacks.writer()
  File "/usr/lib/python2.7/site-packages/pika/diagnostic_utils.py", line 53, in log_exception_func_wrap
    return func(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/pika/adapters/utils/io_services_utils.py", line 1093, in _on_socket_writable
    'but _tx_buffers is empty.', self._state)
AssertionError: ('_AsyncPlaintextTransport._on_socket_writable() called, but _tx_buffers is empty.', 1)
Exception from AsyncMsgManager.run: ('_AsyncPlaintextTransport._on_socket_writable() called, but _tx_buffers is empty.', 1)

## Stop client, sending a Basic.Cancel RPC command to RabbitMQ

PIKA: 2020-10-08 17:59:35,408 - 47455081346816 - pika.channel:241 - DEBUG - Cancelling consumer: ctag1.3b75fac4298444fca8877e52379a3dd4 (nowait=False)
```
Of particular interest are the logs immediately following the publish step which shows a thread enabling the writer interleaved with the ioloop thread.
I have tracked this down to 2 locations in the code.  I made a small change in each, and we're currently doing further testing.

io_services_utils.py around line 1030, small change to handle _tx_buffers empty assertion error:

```
        # NOTE: Modify code to write data to buffer before setting writer.
        # Otherwise a race condition can occur where ioloop executes writer 
        # while buffer is still empty. 
        setWriter = True if not self.get_write_buffer_size() else False

        self._buffer_tx_data(data)

        if setWriter:
            self._nbio.set_writer(self._sock.fileno(), self._on_socket_writable)
            _LOGGER.debug('Turned on writability watcher: %s', self._sock)
```

selector_ioloop_adapter.py around line 326, small change to set the writer callback before enabling the writable mask.

```
            if callbacks.writer is None:
                assert callbacks.reader is not None
                # NOTE: Set the writer func before setting the mask!
                # Otherwise a race condition can occur where ioloop tries to
                # call writer when it is still None.
                callbacks.writer = on_writable
                self._loop.update_handler(
                    fd, self._readable_mask | self._writable_mask)
                LOGGER.debug('set_writer(%s, _) updated handler RdWr', fd)
            else:
                LOGGER.debug('set_writer(%s, _) replacing writer', fd)
                callbacks.writer = on_writable
```

Since we are only using the SelectConnection object, I'm not sure if similar changes would also be needed elsewhere as I'm not familiar enough with the rest of the Pika codebase.

Do you prefer these changes submitted as part of a pull request?  Or are they simple enough that you can add them as part of normal development?

Thanks!

--- a/pika/adapters/utils/io_services_utils.py
+++ b/pika/adapters/utils/io_services_utils.py
@@ -1018,6 +1018,10 @@ class _AsyncPlaintextTransport(_AsyncTransportBase):
         assert data, ('_AsyncPlaintextTransport.write(): empty data from user.',
                       data, self._state)
 
+        # pika/pika#1286
+        # NOTE: Modify code to write data to buffer before setting writer.
+        # Otherwise a race condition can occur where ioloop executes writer 
+        # while buffer is still empty. 
         tx_buffer_was_empty = self.get_write_buffer_size() == 0
 
         self._buffer_tx_data(data)

--- a/pika/adapters/utils/io_services_utils.py
+++ b/pika/adapters/utils/io_services_utils.py
@@ -1160,6 +1160,10 @@ class _AsyncSSLTransport(_AsyncTransportBase):
         assert data, ('_AsyncSSLTransport.write(): empty data from user.',
                       data, self._state)
 
+        # pika/pika#1286
+        # NOTE: Modify code to write data to buffer before setting writer.
+        # Otherwise a race condition can occur where ioloop executes writer 
+        # while buffer is still empty. 
         tx_buffer_was_empty = self.get_write_buffer_size() == 0
 
         self._buffer_tx_data(data)

--- a/pika/adapters/utils/selector_ioloop_adapter.py
+++ b/pika/adapters/utils/selector_ioloop_adapter.py
@@ -325,13 +325,16 @@ class SelectorIOServicesAdapter(io_services_utils.SocketConnectionMixin,
         else:
             if callbacks.writer is None:
                 assert callbacks.reader is not None
+                # NOTE: Set the writer func before setting the mask!
+                # Otherwise a race condition can occur where ioloop tries to
+                # call writer when it is still None.
+                callbacks.writer = on_writable
                 self._loop.update_handler(
                     fd, self._readable_mask | self._writable_mask)
                 LOGGER.debug('set_writer(%s, _) updated handler RdWr', fd)
             else:
                 LOGGER.debug('set_writer(%s, _) replacing writer', fd)
-
-            callbacks.writer = on_writable
+                callbacks.writer = on_writable
 
     def remove_writer(self, fd):
         """Implement

